File: app/src/main/java/art/coded/wireframe/ui/list/ListAdapter.java
package art.coded.wireframe.ui.list;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.app.ShareCompat;
import androidx.recyclerview.widget.RecyclerView;

import art.coded.wireframe.R;
import art.coded.wireframe.data.Element;

import java.util.List;

public class ListAdapter extends RecyclerView.Adapter<ListAdapter.ElementViewHolder> {

    static final String LOG_TAG = ListAdapter.class.getSimpleName();

    LayoutInflater mLayoutInflater;
    List<Element> mAllElements;
    Activity mActivity;

    public ListAdapter(Activity activity) {

        mActivity = activity;
        mLayoutInflater = LayoutInflater.from(activity);
    }

    @NonNull @Override  public ElementViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View itemView = mLayoutInflater.inflate(R.layout.list_tem, parent, false);
        return new ElementViewHolder(itemView, mActivity);
    }

    @Override  public void onBindViewHolder(@NonNull ElementViewHolder holder, int position) {

        if (mAllElements != null) {

            Element element = mAllElements.get(position);
            holder.mItemView.setText(element == null ? "No elements" : element.getElement());
        }
    }

    @Override  public int getItemCount() {

        return mAllElements == null ? 0 : mAllElements.size();
    }

    void setElements(List<Element> elements) {
        mAllElements = elements;
        notifyDataSetChanged();
    }

    Element getElementByPosition(int position) { return mAllElements.get(position); }

    static class ElementViewHolder extends RecyclerView.ViewHolder {

        private final TextView mItemView;
        private final ImageButton mImageButton;

        public ElementViewHolder(@NonNull View itemView, Activity activity) {

            super(itemView);
            mItemView = itemView.findViewById(R.id.textView);
            mImageButton = itemView.findViewById(R.id.list_button_share);
            mImageButton.setOnClickListener(view -> {

                ShareCompat.IntentBuilder
                        .from(activity)
                        .setType("text/plain")
                        .setText(mItemView.getText())
                        .startChooser();
            });
        }
    }
}

File: app/src/ktx/java/art/coded/wireframe/view/adapter/ListAdapter.kt
package art.coded.wireframe.view.adapter

import android.app.Activity
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import art.coded.wireframe.R
import android.widget.TextView
import android.widget.ImageButton
import androidx.core.app.ShareCompat
import android.view.View
import art.coded.wireframe.model.entity.Element

private val LOG_TAG = ListAdapter::class.java.simpleName

class ListAdapter(val activity: Activity?) :
    RecyclerView.Adapter<ListAdapter.ElementViewHolder>() {
    private var layoutInflater: LayoutInflater = LayoutInflater.from(activity)
    private var allElements: List<Element>? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ElementViewHolder {
        val itemView = layoutInflater.inflate(R.layout.list_item, parent, false)
        return ElementViewHolder(itemView, activity)
    }

    override fun onBindViewHolder(holder: ElementViewHolder, position: Int) {
        if (allElements != null) {
            val element = allElements!![position]
            holder.textView.text = element.name ?: "No elements"
        }
    }

    override fun getItemCount(): Int {
        return if (allElements == null) 0 else allElements!!.size
    }

    fun setElements(elements: List<Element>) {
        allElements = elements
        notifyDataSetChanged()
    }

    fun getNameByPosition(position: Int): Element {
        return allElements!![position]
    }

    class ElementViewHolder(itemView: View, activity: Activity?) :
        RecyclerView.ViewHolder(itemView) {
        val textView: TextView = itemView.findViewById(R.id.textView)
        private val imageButton: ImageButton = itemView.findViewById(R.id.list_button_share)

        init {
            imageButton.setOnClickListener { view: View ->
                ShareCompat.IntentBuilder
                    .from(activity!!)
                    .setType("text/plain")
                    .setText(textView.text)
                    .startChooser()
            }
        }
    }
}